[ConstraintSystem] Fix a bogus missing function call error message.

hborla opened the pull request:
If there is no function expression written in the source code, telling the user to insert an explicit function call is really misleading. This can happen when attempting to match the type of a property with attached property wrapper against the `wrappedValue` type:

```swift
@propertyWrapper
struct StringWrappedValue {
  var wrappedValue: String
}

struct SR_14730 {
  @StringWrappedValue var value: () -> String
}
```

Previously, the constraint system attempted to diagnose a missing function call because the result of the function type `() -> String` matches the wrapped-value type `String`, but this isn't the right fix because there's no expression written in the code to actually call (not to mention the diagnostics code not knowing how to handle this case and crashing!).

Resolves: rdar://79051798 / [SR-14730](https://bugs.swift.org/browse/SR-14730)

hborla committed:
[ConstraintSystem] Handle simplifying function result locator path
elements when the anchor is an unapplied decl reference in
simplifyLocator.

In this case, there's nothing to extract, so simply remove the path
element and continue on.

hborla committed:
[ConstraintSystem] Don't attempt missing function call diagnostics if
there is no function expression explicitly written in the source code.

hborla commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -2677,6 +2677,12 @@ void ContextualFailure::tryFixIts(InFlightDiagnostic &diagnostic) const {
+  // isn't written explicitly in the source code.
+  auto *anchor = getAsExpr(simplifyLocatorToAnchor(getLocator()));
+  if (!anchor || anchor->isImplicit())
+    return false;
```
It's annoying that there's a bit of duplicated code here. I wonder if we should always try to diagnose this through `InsertExplicitCall`?

xedin replied:
Yes, that would be great. I think that this is just a leftover from the days of CSDiag, most likely not all the locations have been refactored to produce a tailored `InsertExplicitCall` instead of `IgnoreContextualMismatch` but it was shadowed by the fact that they produce the same diagnostic. Maybe you should just try deleting this whole method and see whether it would be possible to produce a tailored fix where appropriate?

hborla replied:
✅

hborla committed:
[Diagnostics] Diagnose associated value pattern matching errors separately
from missing function call errors in `ContextualFailure`.

It doesn't make sense to diagnose pattern matching errors inside of
`diagnoseMissingFunctionCall`, and the code I added to bail out if there
is no explicit function expression broke the pattern matching diagnostics
anyway.

hborla committed:
[Diagnostics] Remove ContextualFailure::diagnoseMissingFunctionCall.

Instead, always diagnose missing call errors via MissingCallFailure.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test Windows platform

hborla commented:
@swift-ci please test Windows platform
